File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    adj_list = [[] for _ in range(n+1)]
    for a,b in edge:
        adj_list[a].append(b)
        adj_list[b].append(a)
    
    visited = [0]*(n+1)
    
    visited[1]=1
    q = deque([1])
    
    while q:
        cur_node = q.popleft()
        for next_node in adj_list[cur_node]:
            if visited[next_node]==0:
                visited[next_node]= visited[cur_node]+1
                q.append(next_node)

    return visited.count(max(visited))

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    int ans = 0;
    vector<vector<int>> graph(n);
    for(int i=0;i<edge.size();i++){
        int a,b;
        a = edge[i][0];
        b = edge[i][1];
        a--;
        b--;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }
    
    queue<pair<int,int>> q;
    vector<int> visited(n,-1);
    q.push({0,0});
    visited[0]=1;
    while (!q.empty()){
        int node, cnt;
        node = q.front().first;
        cnt = q.front().second;
        q.pop();
      
        for(int i=0;i<graph[node].size();i++){
       
            int nextNode=graph[node][i];
            if (visited[nextNode]==-1){
                visited[nextNode]=cnt+1;
                q.push({nextNode, cnt+1});
            }
        }
    }
    
    for(int i=0;i<n;i++){
        ans = max(ans,visited[i]);
        // cout<<visited[i]<<' ';
    }
    int cnt=0;
    for(int i=0;i<n;i++){
        if(ans==visited[i])cnt++;
    }
    
    return cnt;
}